### PR TITLE
[refactor] Structure configuration files into classes 

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

This PR implements the changes described in this [design doc](https://docs.google.com/document/d/1dTp5y1MRqpgsR2XfgABSId3hcX_pVNiklchjJDi1EXo/edit?ts=5ea748cf#heading=h.v7t1ynb2u4tr). In particular, it replaces the configuration dictionary (called `trainer_params` in many files) with objects based on the `attrs` library. These are more flexible than NamedTuples and support smart defaults, validators, typing, and can be structured/unstructured from Dicts based on the `cattrs` library. 

This PR also assigns reasonable defaults to every hyperparameter, and `mlagents-learn` now allows for training without any configuration file specified (defaults to PPO with extrinsic reward signal). In the future, it will allow for defaults that are being set relative to other defaults (as done in the SelfPlaySettings).  

The core of the changes are in [settings.py](https://github.com/Unity-Technologies/ml-agents/blob/develop-attrs/ml-agents/mlagents/trainers/settings.py). 

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)
Attrs:
https://www.attrs.org/en/stable/

Cattrs:
https://cattrs.readthedocs.io/en/latest/

This PR does not yet refactor how the settings are passed down from Trainer->Policy-> Optimizer. The other two objects still receive the entire `TrainerSettings` object. 
TODO: Improved testing for error handling in `settings.py`, ~Update docs.~ 

NOTE: The appearance of `output_path` as `default` in the configs is a bit odd, as `output_path` is not settable through the configuration (it's determined by the `run-id`). I plan to remove `output_path` in a follow-up PR. 

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [x] Breaking change
- [x] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [x] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [x] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/trainer_util.py [hunk 1]
```diff
@@ -90,27 +90,10 @@ def initialize_trainer(
...
-    if init_path is not None:
-        trainer_parameters["init_path"] = os.path.join(init_path, brain_name)
-    trainer_parameters["keep_checkpoints"] = keep_checkpoints
-    if brain_name in trainer_config:
-        _brain_key: Any = brain_name
-        while not isinstance(trainer_config[_brain_key], dict):
-            _brain_key = trainer_config[_brain_key]
-        trainer_parameters.update(trainer_config[_brain_key])
-
+    print(trainer_config)
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer_util.py:93]: Remove?

File: ml-agents/mlagents/trainers/trainer_util.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def generate(self, brain_name: str) -> Trainer:
 
 
 def initialize_trainer(
-    trainer_config: Any,
+    trainer_config: TrainerSettings,
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer_util.py:63]: minor but.. `trainer_settings` if we are calling the class `TrainerSettings`?

File: ml-agents/mlagents/trainers/trainer_util.py [hunk 3]
```diff
@@ -1,6 +1,5 @@
 import os
-import yaml
-from typing import Any, Dict, TextIO
+from typing import Any, Dict
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer_util.py:2]: The `Any` here refers to the trainer_config param to `TrainerFactory`.  Should that be updated with the new class name?

File: ml-agents/mlagents/trainers/trainer/trainer.py [hunk 4]
```diff
@@ -25,7 +26,7 @@ class Trainer(abc.ABC):
     def __init__(
         self,
         brain_name: str,
-        trainer_parameters: dict,
+        trainer_parameters: TrainerSettings,
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer/trainer.py:29]: trainer_parameters -> trainer_settings?

File: ml-agents/mlagents/trainers/trainer/trainer.py [hunk 5]
```diff
@@ -80,7 +81,7 @@ def get_max_steps(self) -> int:
         Returns the maximum number of steps. Is used to know when the trainer should be stopped.
         :return: The maximum number of steps of the trainer
         """
-        return int(float(self.trainer_parameters["max_steps"]))
+        return int(float(self.trainer_parameters.max_steps))
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer/trainer.py:84]: Does this need the `int(float(_))`?  I think we should be able to manage the type coersion in the parsing logic upstream of this.

File: ml-agents/mlagents/trainers/trainer/rl_trainer.py [hunk 6]
```diff
@@ -40,7 +33,8 @@ def __init__(self, *args, **kwargs):
         }
         self.update_buffer: AgentBuffer = AgentBuffer()
         self._stats_reporter.add_property(
-            StatsPropertyType.HYPERPARAMETERS, self.trainer_parameters
+            StatsPropertyType.HYPERPARAMETERS,
+            cattr.unstructure(self.trainer_parameters),
```
Comment (harperj) [inline ml-agents/mlagents/trainers/trainer/rl_trainer.py:37]: Minor but this feels like it leaks the abstraction of the `TrainerSettings` implementation.  Perhaps this could be implemented as `TrainerSettings.as_dict`?

File: ml-agents/mlagents/trainers/settings.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+    swap_steps: int = 10000
+    window: int = 10
+    play_against_latest_model_ratio: float = 0.5
+    initial_elo: float = 1200.0
+
+
+@attr.s(auto_attribs=True)
+class TrainerSettings:
+    # Edit these two fields to add new trainers #
+    class TrainerType(Enum):
```
Comment (harperj) [inline ml-agents/mlagents/trainers/settings.py:201]: IMO this class and the associated static method should live fully outside of `TrainerSettings`.  

So...

```python
class TrainerType(Enum):
  PPO: str = "ppo"
  SAC: str = "sac"

  def to_settings(self) -> type:
    _mapping = {
      self.PPO: PPOSettings,
      self.SAC: SACSettings,
    }
    return _mapping[self]
```

I also used self-references for the members, which might arguably be less clear.  That isn't necessary.

File: ml-agents/mlagents/trainers/settings.py [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+
+
+@attr.s(auto_attribs=True)
+class BehavioralCloningSettings:
+    demo_path: str
+    steps: int = 0
+    strength: float = 1.0
+    samples_per_update: int = 0
+    num_epoch: Optional[int] = None
+    batch_size: Optional[int] = None
```
Comment (harperj) [inline ml-agents/mlagents/trainers/settings.py:111]: Could you clarify what it means for these to be `None`?  Might be worth a comment.

File: ml-agents/mlagents/trainers/settings.py [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+        except KeyError:
+            raise TrainerConfigError(f"Unknown reward signal type {key}")
+    return d_final
+
+
+def defaultdict_to_dict(d: DefaultDict) -> Dict:
+    return {key: cattr.unstructure(val) for key, val in d.items()}
+
+
+def dict_to_defaultdict(d: Dict, t: type) -> DefaultDict:
```
Comment (harperj) [inline ml-agents/mlagents/trainers/settings.py:84]: Seems like this really is intended for `TrainerSettings` only since `t` is unused.  Maybe you could make it a class method?

File: ml-agents/mlagents/trainers/settings.py [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+            d_copy[key] = int(
+                float(val)
+            )  # In some configs, max steps was specified as a float
+        else:
+            d_copy[key] = check_and_structure(key, val, t)
+    return t(**d_copy)
+
+
+def rewardsignal_settings_to_cls(d: Mapping) -> Any:
+    if d is None:
```
Comment (harperj) [inline ml-agents/mlagents/trainers/settings.py:66]: Should / could we just prevent `None` from being passed here?  If not, `d` should be an `Optional[Mapping]`.

File: ml-agents/mlagents/trainers/settings.py [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+from enum import Enum
+import collections
+import argparse
+
+from mlagents.trainers.cli_utils import StoreConfigFile, DetectDefault, parser
+from mlagents.trainers.cli_utils import load_config
+from mlagents.trainers.exception import TrainerConfigError
+from mlagents.trainers.models import LearningRateSchedule, EncoderType
+
+
```
Comment (harperj) [inline ml-agents/mlagents/trainers/settings.py:13]: In general these parsing helper methods at the top could use a little love, IMO.  They scatter the parsing / error handling a bit.  Ideally there would be a way to embed them closer to the types/fields themselves.  If that's difficult for some reason at least some comments and/or more verbose field names might be helpful

File: ml-agents/mlagents/trainers/trainer/trainer.py (hunk 5 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/trainer/trainer.py]: You're right, this is already enforced during the structuring. Removed.

File: ml-agents/mlagents/trainers/trainer/trainer.py (hunk 4 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/trainer/trainer.py]: Yep! Changed throughout

File: ml-agents/mlagents/trainers/trainer_util.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/trainer_util.py]: 👍

File: ml-agents/mlagents/trainers/trainer/rl_trainer.py (hunk 6 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/trainer/rl_trainer.py]: Made a mixin that defines the `as_dict` method and gave it to TrainerSettings and RunOptions - should clean up the code in learn.py a bit.

File: ml-agents/mlagents/trainers/settings.py (hunk 7 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py]: Split off the type enums (TrainerType and RewardSIgnalType) as separate classes.

File: ml-agents/mlagents/trainers/settings.py (hunk 8 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py:95]: For the BC module, not specifying num_epoch and batch_size means that it will use the current values for these from the trainer. This logic happens in the BC module itself - I couldn't find a good way to put the logic in `settings.py`. Will add a comment

File: ml-agents/mlagents/trainers/settings.py (hunk 10 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py]: True, these should never be None unless we have a malformed config (e.g. if user puts `env_settings: null` in their config). I now throw an error if a None (or non-Mapping) is received.

File: ml-agents/mlagents/trainers/settings.py (hunk 11 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py:13]: I've moved them into their appropriate classes as static methods - the downside being that certain classes are now quite large. But I think it's an improvement over having those methods up at the top.

File: config/upgrade_config.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
+import attr
```
Comment (andrewcoh) [inline config/upgrade_config.py:1]: How long do you think we'll need to keep this script around?

File: docs/Training-Configuration-File.md [hunk 13]
```diff
@@ -26,20 +26,20 @@ choice of the trainer (which we review on subsequent sections).
...
+| `max_steps`              | (default = `500000`) Total number of experience points that must be collected from the simulation before ending the training process. <br><br>Typical range: `5e5` - `1e7`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
```
Comment (andrewcoh) [inline docs/Training-Configuration-File.md:36]: 'experience points'

File: ml-agents/mlagents/trainers/components/bc/module.py [hunk 14]
```diff
@@ -36,14 +31,16 @@ def __init__(
...
+        self.model = BCModel(policy, self.current_lr, settings.steps)
         _, self.demonstration_buffer = demo_to_buffer(
-            demo_path, policy.sequence_length, policy.brain
+            settings.demo_path, policy.sequence_length, policy.brain
         )
 
-        self.batch_size = batch_size if batch_size else default_batch_size
-        self.num_epoch = num_epoch if num_epoch else default_num_epoch
+        self.batch_size = (
+            settings.batch_size if settings.batch_size else default_batch_size
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/components/bc/module.py:41]: why does this need to be a tuple now?

File: ml-agents/mlagents/trainers/components/reward_signals/curiosity/signal.py [hunk 15]
```diff
@@ -26,9 +20,11 @@ def __init__(
         :param encoding_size: The size of the hidden encoding layer for the ICM
```
Comment (andrewcoh) [inline ml-agents/mlagents/trainers/components/reward_signals/curiosity/signal.py:20]: param docstring :(

File: config/upgrade_config.py (hunk 12 above)
Comment (ervteng) [reply config/upgrade_config.py:1]: Hopefully only for 1-2 releases. Or perhaps it should be released with Trainer 1.0 and removed after that.

File: ml-agents/mlagents/trainers/components/bc/module.py (hunk 14 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/components/bc/module.py:41]: It's not a tuple, it's the way `black` likes to format long lines 😛

File: ml-agents/mlagents/trainers/ppo/trainer.py [hunk 16]
```diff
@@ -76,9 +80,9 @@ def _check_param_keys(self):
         # Check that batch size is greater than sequence length. Else, throw
         # an exception.
         if (
-            self.trainer_parameters["sequence_length"]
-            > self.trainer_parameters["batch_size"]
-            and self.trainer_parameters["use_recurrent"]
+            self.trainer_settings["sequence_length"]
+            > self.trainer_settings["batch_size"]
+            and self.trainer_settings["use_recurrent"]
```
Comment (harperj) [inline ml-agents/mlagents/trainers/ppo/trainer.py:85]: I think this can be:
```suggestion
            self.trainer_settings.sequence_length
            > self.trainer_settings.batch_size
            and self.trainer_settings.use_recurrent
```

File: ml-agents/mlagents/trainers/sac/optimizer.py [hunk 17]
```diff
@@ -44,32 +45,35 @@ def __init__(self, policy: TFPolicy, trainer_params: Dict[str, Any]):
...
-                self.tau = trainer_params.get("tau", 0.005)
-                self.burn_in_ratio = float(trainer_params.get("burn_in_ratio", 0.0))
+                self.act_size = policy.act_size
+                policy_network_settings = policy.network_settings
+                h_size = policy_network_settings.hidden_units
+                num_layers = policy_network_settings.num_layers
+                vis_encode_type = policy_network_settings.vis_encode_type
+
+                self.tau = hyperparameters.tau
+                self.burn_in_ratio = 0.0
```
Comment (harperj) [inline ml-agents/mlagents/trainers/sac/optimizer.py:65]: Looks like this used to be a trainer param but is not just being set to 0.0?  Any reason why?

File: ml-agents/mlagents/trainers/sac/trainer.py [hunk 18]
```diff
@@ -47,75 +47,44 @@ def __init__(
...
         super()._check_param_keys()
         # Check that batch size is greater than sequence length. Else, throw
         # an exception.
         if (
-            self.trainer_parameters["sequence_length"]
-            > self.trainer_parameters["batch_size"]
-            and self.trainer_parameters["use_recurrent"]
+            self.trainer_settings["sequence_length"]
+            > self.trainer_settings["batch_size"]
+            and self.trainer_settings["use_recurrent"]
```
Comment (harperj) [inline ml-agents/mlagents/trainers/sac/trainer.py:87]:
```suggestion
            self.trainer_settings.sequence_length
            > self.trainer_settings.batch_size
            and self.trainer_settings.use_recurrent
```

File: ml-agents/mlagents/trainers/settings.py [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+    def _set_default_hyperparameters(self):
+        return self.trainer_type.to_settings()()
+
+    network_settings: NetworkSettings = attr.ib(factory=NetworkSettings)
+    reward_signals: Dict[RewardSignalType, RewardSignalSettings] = attr.ib(
+        factory=lambda: {RewardSignalType.EXTRINSIC: RewardSignalSettings()}
+    )
+    init_path: Optional[str] = None
+    output_path: str = "default"
+    # TODO: Remove parser default and remove from CLI
```
Comment (harperj) [inline ml-agents/mlagents/trainers/settings.py:194]: Address before merging?

File: ml-agents/mlagents/trainers/sac/trainer.py (hunk 18 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/sac/trainer.py]: Ah, this method isn't actually used anymore, the check is done in `settings.py` in `_check_batch_size_seq_length`. I've removed them from PPO and SAC.

File: ml-agents/mlagents/trainers/ppo/trainer.py (hunk 16 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/ppo/trainer.py]: Ah, this method isn't actually used anymore, the check is done in `settings.py` in `_check_batch_size_seq_length`. I've removed them from PPO and SAC.

File: ml-agents/mlagents/trainers/sac/optimizer.py (hunk 17 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/sac/optimizer.py:65]: This wasn't configurable before (also set to 0 as a constant at the top of the file). Currently the burn-in feature doesn't actually work properly with values greater than 0. Will fix or remove in a PR that just touches that.

File: ml-agents/mlagents/trainers/settings.py (hunk 19 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/settings.py]: Done!
